tests-ng/plugins/sysdiff.py

```diff
@@ -0,0 +1,482 @@
+"""
+System state snapshot and diff functionality for tests-ng.
+
+Provides snapshot and comparison capabilities for:
+- packages (dpkg)
+- systemd units
+- files (sha256 checksums)
+- sysctl parameters
+- loaded kernel modules
+"""
+
+import difflib
+import json
+import os
+import re
+import tempfile
+from dataclasses import dataclass
+from datetime import datetime
+from pathlib import Path
+from typing import Dict, List, Optional, Tuple
+
+import pytest
+
+from .shell import ShellRunner
+from .dpkg import Dpkg
+from .systemd import Systemd, SystemdUnit
+from .kernel_module import KernelModule
+from .sysctl import Sysctl
+
+
+DEFAULT_PATHS = [
+    "/etc", "/boot", "/usr/local/bin", "/usr/local/sbin",
+    "/usr/local/lib", "/opt", "/proc/mounts"
+]
+
+STATE_DIR = Path(os.environ.get("XDG_STATE_HOME", Path.home() / ".local/state")) / "sysdiff"
+
+IGNORED_SYSTEMD_PATTERNS = []
+IGNORED_KERNEL_MODULES = []
+
+
+@dataclass
+class SnapshotMetadata:
+    """Metadata for a snapshot"""
+    created_at: str
+    paths: List[str]
+    ignore_file: bool
+
+
+@dataclass
+class Snapshot:
+    """Complete system snapshot"""
+    name: str
+    metadata: SnapshotMetadata
+    packages: Dict[str, str]  # package -> version
```

Comment:
any reason for not using data classes for packages, files, systcl params instead of dict[str, str]? this would be easier to read and safer. tbh, I'm not sure if this then could also be used with a dict or if it would have to be a list, but I assume this would not be too bad with the amount of data we have here and readability/safety is in my view the more important aspect than efficiency

Reply:
Having dataclasses for everything is way more consistent, right. Every class that we use to gather data now has a propper dataclass.